### ISSUE TITLE
Google AMP cache issue: randomly redirects to non-cached version

### ISSUE THREAD

**Issue by omerr132** [ampproject/amphtml#36471]
### Description

The following issue takes place on multiple sites we manage:

We create AMP pages, and use paired mode to associate them with the HTML page. The AMP pages are valid.
According to the search console, Google identifies and indexes them correctly.

However, when accessing the cached AMP pages from the Google search results, as well as directly via the cached URL, the page redirects to the non-cached version on our server. 

The issue seems to be happening irregularly across different pages of the same site and occasionally between different loadings of the same page - sometimes no redirection takes place and sometimes it does. 

For example, the following page's cached version redirects to the non-cached version (both when loaded directly, or via the search results):

- **Source HTML**: http://www.pacehospital.com/irritable-bowel-syndrome-ibs-dos-and-dont-s 
- **AMP page**: http://www.pacehospital.com/irritable-bowel-syndrome-ibs-dos-and-dont-s?amp=true
- **Cached version**: https://www-pacehospital-com.cdn.ampproject.org/c/www.pacehospital.com/irritable-bowel-syndrome-ibs-dos-and-dont-s?amp=true
 
If it helps - the results from Google AMP Cache URL API are also irregular and aren't always in-sync with the search console, or the actual cache behavior. In this case however, the API seems to identify the HTML URL as the AMP URL (i.e. the AMP URL is missing the _?amp=true_ parameter):

{
  "ampUrls": [
    {
      "originalUrl": "http://www.pacehospital.com/irritable-bowel-syndrome-ibs-dos-and-dont-s",
      "ampUrl": "http://www.pacehospital.com/irritable-bowel-syndrome-ibs-dos-and-dont-s",
      "cdnAmpUrl": "https://www-pacehospital-com.cdn.ampproject.org/c/www.pacehospital.com/irritable-bowel-syndrome-ibs-dos-and-dont-s"
    }
  ]
}

We use JS redirect in the AMP page due to restrictions in the CMS. As far as we know, it is allowed. And indeed it works most of the times. 

This bug is similar in nature to https://github.com/ampproject/amphtml/issues/35262, but I believe it is not the same bug.

### Reproduction Steps

1.  Verify the AMP is valid: https://search.google.com/test/amp/result?id=3AU4LEeg3QMbFhEibHuF3g&has_paired_amp=true

2. Open the cached version of the page: https://www-pacehospital-com.cdn.ampproject.org/c/www.pacehospital.com/irritable-bowel-syndrome-ibs-dos-and-dont-s?amp=true

3. Alternatively, open the cached AMP page via the mobile search results: https://www.google.com/search?q=pacehospital+com+irritable-bowel-syndrome-ibs-dos-and-dont-s&oq=pacehospital+com+irritable-bowel-syndrome-ibs-dos-and-dont-s&aqs=chrome..69i57j69i60l3.5845j0j4&sourceid=chrome&ie=UTF-8

You'll notice the page has redirected from the AMP cached version to the AMP url (which then has JS redirect to the actual non-cached AMP page on the convert.ampify.io domain).

No errors are logged in the dev console. 

### Browser(s) Affected

Chrome, Safari

**Comment by omerr132** [comment]
Any thoughts or questions?

**Comment by alanorozco** [comment]
@ampproject/wg-caching

**Comment by Gregable** [comment]
```
$ curl -i https://www-pacehospital-com.cdn.ampproject.org/c/www.pacehospital.com/irritable-bowel-syndrome-ibs-dos-and-dont-s?amp=true
HTTP/2 200 
location: http://www.pacehospital.com/irritable-bowel-syndrome-ibs-dos-and-dont-s?amp=true
cache-control: private
x-silent-redirect: true
warning: 199 - "Failed to validate PCU."
```

The `warning:` header here explains the redirect reason. When the cache crawled the page, it was invalid. The irregularity is probably because the origin page goes valid/invalid over time. 

You can run some tests on your website to see what's going on. First, let's set a variable:

```
$ TESTURL="https://www.pacehospital.com/irritable-bowel-syndrome-ibs-dos-and-dont-s?amp=true"
```

Then, let's see what your server is returning:

```
curl $TESTURL | less

<!doctype html >
<html xmlns="http://www.w3.org/1999/xhtml" lang="en"
      class="">
<head>
<script type="text/javascript">
    window._currentDevice = 'desktop';
...
```

Which isn't valid AMP.

We can try requesting with a few different user agents, the first one is what https://validator.amp.dev/ tries to use:

```
AMPDEV="Mozilla/5.0 (Linux; Android 6.0.1; Nexus 5X Build/MTC19V) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.81 Mobile Safari/537.36 (compatible; validator.ampproject.org) AppEngine-Google; (+http://code.google.com/appengine; appid: s~amp-validator)"
GOOGLEBOT="Mozilla/5.0 (Linux; Android 6.0.1; Nexus 5X Build/MMB29P) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.100 Mobile Safari/537.36 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)"
DESKTOP="Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.87 Safari/537.36"
MOBILE="Mozilla/5.0 (Linux; Android 7.1.1; Nexus 6 Build/N6F26U) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.100 Mobile Safari/537.36"

curl --silent -H 'user-agent: $AMPDEV' $TESTURL | md5sum
curl --silent -H 'user-agent: $DESKTOP' $TESTURL | md5sum
curl --silent -H 'user-agent: $MOBILE' $TESTURL | md5sum
curl --silent -H 'user-agent: $GOOGLEBOT' $TESTURL | md5sum

bf746e145dc8547613240d463c32a8cb  -
bf746e145dc8547613240d463c32a8cb  -
815a8a64ef71217b2b5e307d77d60b28  -
bf746e145dc8547613240d463c32a8cb  -
```

It seems the server responds differently to a mobile useragent than the other useragents. If I look at what the mobile useragent is returning, I see:

```
$ curl -H 'user-agent: $MOBILE' $TESTURL | less

<!doctype html >
<html xmlns="http://www.w3.org/1999/xhtml" lang="en"
      class="">
<head>
<script type="text/javascript">
    window._currentDevice = 'desktop';
...
```

Still not valid amp though. However, if I run the command multiple times, I get a different response each time:

```
$ curl --silent -H 'user-agent: $MOBILE' $TESTURL | md5sum
a9fd6597f0596529550a27ef9b647218  -
$ curl --silent -H 'user-agent: $MOBILE' $TESTURL | md5sum
34f7c891b8d8ce9472f6c83a751d345d  -
$ curl --silent -H 'user-agent: $MOBILE' $TESTURL | md5sum
eb06aaa7256dee6dd268b679b84eb05e  -
$ curl --silent -H 'user-agent: $MOBILE' $TESTURL | md5sum
2027999ef287f1999396a64ca2b94931  -
$ curl --silent -H 'user-agent: $MOBILE' $TESTURL | md5sum
cd35d4511b347f6ab24316061e7e8daa  -
```

So maybe sometimes your server returns AMP and sometimes it returns something else.

You can also try this by installing the [AMP Validator Node Tool](https://www.npmjs.com/package/amphtml-validator), and then running something like:

```
$ curl -H 'user-agent: $MOBILE' $TESTURL | amphtml-validator /dev/stdin
/dev/stdin:2:0 The mandatory attribute '⚡' is missing in tag 'html'. (see https://amp.dev/documentation/guides-and-tutorials/learn/spec/amphtml#required-markup)
/dev/stdin:2:0 The attribute 'xmlns' may not appear in tag 'html'. (see https://amp.dev/documentation/guides-and-tutorials/learn/spec/amphtml/#required-markup)
/dev/stdin:42:0 Custom JavaScript is not allowed. (see https://amp.dev/documentation/guides-and-tutorials/learn/validation-workflow/validation_errors/#custom-javascript-is-not-allowed)
/dev/stdin:165:0 Custom JavaScript is not allowed. (see https://amp.dev/documentation/guides-and-tutorials/learn/validation-workflow/validation_errors/#custom-javascript-is-not-allowed)
/dev/stdin:174:0 Custom JavaScript is not allowed. (see https://amp.dev/documentation/guides-and-tutorials/learn/validation-workflow/validation_errors/#custom-javascript-is-not-allowed)
...
```

For me, this shows hundreds of validation errors.

tl;dr: The root issue is that your web server appears to be returning non-amp pages, which don't validate. It may be the case that the server sometimes returns valid pages and this flips back and forth over time. Some of the above debugging steps can hopefully help you debug further.
